{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from os import chdir\n",
    "import subprocess\n",
    "import glob\n",
    "\n",
    "# local import (Hirotaka's myfunctions.py script)\n",
    "import myfunctions as myfunc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory 'temp': File exists\n"
     ]
    }
   ],
   "source": [
    "!mkdir temp\n",
    "chdir('temp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the id file, covariate file (all European)\n",
    "wgsfolder = '/data/CARD/PD/GENOMES/august19/genotypes'\n",
    "idfile = '_keep.id'\n",
    "demog = pd.read_csv('/data/LNG/iwakih2/ampPD/data1_out/PP_PD_BF_HB_LC_demog.csv')\n",
    "euro = pd.read_csv(f'{wgsfolder}/PCA_filtered_europeans.txt', sep=' ',\n",
    "                  header=None, names=['FID', 'IID'])\n",
    "df = pd.merge(demog, euro, left_on='WGSID', right_on='IID')\n",
    "df['FID'] = df.WGSID\n",
    "df['IID'] = df.WGSID\n",
    "df.RECRUIT=['CTR' if i=='HC' else i for i in df.RECRUIT]\n",
    "df = df[df.RECRUIT.isin(['CTR', 'PD'])]\n",
    "df = df[df.RECRUIT==df.LASTDIAG]\n",
    "df['pheno'] = [1 if i=='PD' else 0 for i in df.RECRUIT]\n",
    "df['AAD'] = df.AADi.where(df.pheno==0, other=df.AAOi)\n",
    "df['AGE'] = df.AGEatBL.where(df.pheno==0, other=df.AAD)\n",
    "\n",
    "df[['FID', 'IID', 'pheno']].to_csv(idfile, index=False, sep='\\t', header=None)\n",
    "df[['FID', 'IID', 'AGE', 'FEMALE']].to_csv('_qcovar.txt', index=False, sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for chrnum in range(1,23):\n",
    "#     pfile = f'{wgsfolder}/pd.june2019.chr{chrnum}.sqc'\n",
    "#     out = f'{wgsfolder}/merge/pd.june2019.chr{chrnum}.sqc'\n",
    "#     plinkcmd = f'\\\n",
    "#  plink2 --pfile {pfile}\\\n",
    "#  --make-bed\\\n",
    "#  --out {out}'\n",
    "#    myfunc.shell_do(plinkcmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open('_mergelist.txt', 'w') as f:\n",
    "#     for chrnum in range(1,23):\n",
    "#         bfile = f'{wgsfolder}/merge/pd.june2019.chr{chrnum}.sqc'\n",
    "#         f.write(f'{bfile}\\n')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mergecmd = f'\\\n",
    "# plink --merge-list _mergelist.txt\\\n",
    "#  --make-bed\\\n",
    "#  --out {wgsfolder}/merge/pd.june2019.sqc'\n",
    "# with open('_merge_batch_script.sh', 'w') as f:\n",
    "#     f.write('#!/bin/bash\\n')\n",
    "#     f.write('module load plink\\n')\n",
    "#     f.write(f'{mergecmd}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !sbatch _merge_batch_script.sh --cpus-per-task=128 --mem=100g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# used this to generate pca for chr1 just for testing purposes-- prune and merge all chrs first then run PCA\n",
    "with open('_pca.swarm', 'w') as f:\n",
    "    for chrnum in range(1,2):\n",
    "        pfile = f'{wgsfolder}/pd.june2019.chr{chrnum}.sqc'\n",
    "        out_pcs = f'chr{chrnum}.pca'\n",
    "        plink_pca_cmd = f'\\\n",
    "plink2 --pfile {pfile}\\\n",
    " --pca 50\\\n",
    " --out {out_pcs}\\n'\n",
    "        \n",
    "        f.write(plink_pca_cmd)\n",
    "        print(plink_pca_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !swarm -f _pca.swarm -g 64 --time=10:00:00 -t 20 --logdir swarm --module plink/2.0_alpha_1_final --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('_script1.sh', 'w') as f:\n",
    "    for chrnum in range(1,):\n",
    "        pfile = f'{wgsfolder}/pd.june2019.chr{chrnum}.sqc'\n",
    "        freqfile = f'freq.{chrnum}'\n",
    "        plinkcmd = f\"\\\n",
    "plink2 --pfile {pfile}\\\n",
    " --keep {idfile} \\\n",
    " --freq \\\n",
    " --out {freqfile}\\n\"\n",
    "        f.write(plinkcmd)\n",
    "#         myfunc.shell_do(plinkcmd)\n",
    "print(f'{plinkcmd}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dask.distributed import Client, progress\n",
    "# client should match the qued number of threads\n",
    "# client = Client(processes=False, threads_per_worker=2,\n",
    "#                 n_workers=2, memory_limit='30GB')\n",
    "# import dask\n",
    "import dask.dataframe as dd\n",
    "\n",
    "#try testing fewer maf bins\n",
    "def MAFbin(x):\n",
    "    if x>0:\n",
    "        if x<0.001:\n",
    "            return(1)\n",
    "        elif x<0.01:\n",
    "            return(2)\n",
    "        elif x<0.1:\n",
    "            return(3)\n",
    "        elif x<0.25:\n",
    "            return(4)\n",
    "        elif x<0.5:\n",
    "            return(5)\n",
    "        else:\n",
    "            return(6)\n",
    "    else:\n",
    "        return(0)\n",
    "\n",
    "\n",
    "\n",
    "# def MAFbin(x):\n",
    "#     if x>0:\n",
    "#         if x<0.001:\n",
    "#             return(1)\n",
    "#         elif x<0.01:\n",
    "#             return(2)\n",
    "#         elif x<0.1:\n",
    "#             return(3)\n",
    "#         elif x<0.2:\n",
    "#             return(4)\n",
    "#         elif x<0.3:\n",
    "#             return(5)\n",
    "#         elif x<0.4:\n",
    "#             return(6)\n",
    "#         else:\n",
    "#             return(7)\n",
    "#     else:\n",
    "#         return(0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "client.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_counts_df = pd.DataFrame(index=[0,1,2,3,4,5,6])\n",
    "\n",
    "for chrnum in range(1,23):\n",
    "    freqfile = f'freq.{chrnum}'\n",
    "\n",
    "    df = dd.read_csv(f'{freqfile}.afreq', sep='\\t')\n",
    "    df['MAF'] = df.ALT_FREQS.where(cond=df.ALT_FREQS<=0.5, other=1-df.ALT_FREQS)\n",
    "    df['Group'] = df.MAF.apply(MAFbin, meta=('MAF', 'float64'))\n",
    "    for i in range(1,7):\n",
    "        df.loc[df.Group==i, ['ID']].to_csv(\n",
    "            f'{freqfile}.mafbin{i}', index=False, single_file=True)\n",
    "    print(f'chr{chrnum} finished')\n",
    "    total_counts_df[chrnum] = pd.DataFrame(df.groupby('Group').ID.count().compute())\n",
    "#     total_counts_df.merge(counts_df, left_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # add chr names\n",
    "chrs = ['chr' + str(i) for i in range(1,23)]\n",
    "total_counts_df.columns = chrs\n",
    "\n",
    "# add bins\n",
    "bins = ['0.000', '<0.001', '<0.01', '<0.1', '<0.25', '<0.5', '=0.5']\n",
    "total_counts_df['bin'] = bins\n",
    "total_counts_df.set_index('bin', inplace=True)\n",
    "total_counts_df['total'] = total_counts_df.sum(axis=1)\n",
    "total_counts_df.to_csv('_maf_bin_counts.txt', sep='\\t')\n",
    "total_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "freqfile = f'freq.21'\n",
    "\n",
    "df = pd.read_csv(f'{freqfile}.afreq', sep='\\t')\n",
    "df['MAF'] = df.ALT_FREQS.where(cond=df.ALT_FREQS<=0.5, other=1-df.ALT_FREQS)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Group'] = df.MAF.apply(MAFbin)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df.Group == 6].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('_script2.swarm', 'w') as f:\n",
    "    for chrnum in range(1,23):\n",
    "        pfile = f'{wgsfolder}/pd.june2019.chr{chrnum}.sqc'\n",
    "        freqfile = f'freq.{chrnum}'\n",
    "        for Indx in range(1,7):\n",
    "            mafbinIndx = f'mafbin{Indx}'\n",
    "            mafbinfile =  f'{freqfile}.{mafbinIndx}'\n",
    "\n",
    "            plinkcmd = f\"\\\n",
    "plink2 --pfile {pfile}\\\n",
    " --extract {mafbinfile}\\\n",
    " --keep {idfile}\\\n",
    " --make-bed\\\n",
    " --out {mafbinfile}\"\n",
    "            gctacmd1 = f'\\\n",
    "gcta64 --bfile {mafbinfile}\\\n",
    " --ld-score-region 10000\\\n",
    " --out {mafbinfile}.ldbin\\\n",
    " --thread-num 10'\n",
    "\n",
    "            f.write(f'{plinkcmd} && {gctacmd1}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !swarm -f _script2.swarm --time=40:00:00 -g 64 -t 10 --logdir swarm --module GCTA,plink/2.0_alpha_1_final --partition=norm --devel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for chrnum in range(1,23):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to rerun for jobs that were killed\n",
    "outputs = glob.glob('freq.*.mafbin*.ldbin.score.ld')\n",
    "# len(glob.glob('*.ldbin.log'))\n",
    "expected_out = ['freq.' + str(i) + '.mafbin' + str(j) + '.ldbin.score.ld' for i in range(1,23) for j in range(1,7)]\n",
    "missing = [bed.replace('.ldbin.score.ld', '') for bed in list(set(expected_out) - set(outputs))]\n",
    "\n",
    "with open('_rerunscript.swarm', 'w') as f:\n",
    "    for mafbinfile in missing:\n",
    "        gctacmd1 = f'\\\n",
    "gcta64 --bfile {mafbinfile}\\\n",
    " --ld-score-region 10000\\\n",
    " --out {mafbinfile}.ldbin\\\n",
    " --thread-num 10'\n",
    "\n",
    "        f.write(f'{gctacmd1}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!swarm -f _rerunscript.swarm --time=40:00:00 -g 64 -t 10 --logdir swarm --module GCTA --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for chrnum in range(1,23):\n",
    "    pfile = f'{wgsfolder}/pd.june2019.chr{chrnum}.sqc'\n",
    "    freqfile = f'freq.{chrnum}'\n",
    "    for Indx in range(1,7):\n",
    "        mafbinIndx = f'mafbin{Indx}'\n",
    "        mafbinfile =  f'{freqfile}.{mafbinIndx}'\n",
    "\n",
    "        \n",
    "        \n",
    "\n",
    "        # Note this is the ldscore within the maf bin\n",
    "        df = dd.read_csv(f'{mafbinfile}.ldbin.score.ld', sep=' ')\n",
    "        ldscvec = df.ldscore_SNP.compute()\n",
    "        ldscThres = ldscvec.median()\n",
    "        df['Group'] = 1\n",
    "        df['Group'] = df.Group.where(cond=df.ldscore_SNP<ldscThres, other=2)\n",
    "\n",
    "\n",
    "        # Create ldbin and GRMs within the ldbin\n",
    "        for i in range(1, 3):\n",
    "            ldbinIndx = f'{mafbinfile}.ldbin{i}'\n",
    "#             df.loc[df.Group==i, ['SNP']].to_csv(\n",
    "#                 f'{ldbinIndx}', index=False, single_file=True)\n",
    "#             gctacmd2 = f'\\\n",
    "#             gcta64 --bfile {mafbinfile}\\\n",
    "#  --extract {ldbinIndx}\\\n",
    "#  --make-grm\\\n",
    "#  --out {ldbinIndx}'\n",
    "#             myfunc.shell_do(gctacmd2)\n",
    "            with open('multi_GRMs.txt', 'a') as f:\n",
    "                f.write(f'{ldbinIndx}\\n')\n",
    "                print(ldbinIndx)\n",
    "        # mafbin information save to mafbinInfo.csv\n",
    "        with open('mafbinInfo.txt', 'a') as f:\n",
    "            s = f'{chrnum}\\t{mafbinIndx}\\t{len(ldscvec)}\\t{ldscvec.mean()}\\t{ldscvec.median()}\\t{ldscvec.std()}'\n",
    "            f.write(f'{s}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create covar file per chrom and impute missing age (may want to consider dropping age column in this case (missing for more than half))\n",
    "# impute \"FEMALE\" with median\n",
    "for chrnum in range(1,2):\n",
    "    pcs_df = pd.read_csv(f'chr{chrnum}.pca.eigenvec', sep='\\t')\n",
    "    pcs_df.rename(columns={'#FID':'FID'}, inplace=True)\n",
    "    covar_df = df[['FID', 'IID', 'AGE', 'FEMALE']].merge(pcs_df, on=['FID','IID'])\n",
    "    covar_df.drop(['AGE','FEMALE'], axis=1, inplace=True)\n",
    "#     covar_df.AGE.fillna(covar_df.AGE.mean(), inplace=True)\n",
    "#     covar_df.FEMALE.fillna(covar_df.FEMALE.median(),inplace=True)\n",
    "#     print(f'dropping {covar_df[covar_df.AGE.isnull()].shape[0]} rows for missing age')\n",
    "#     covar_df.dropna(how='any', inplace=True)\n",
    "    covar_df.to_csv(f'chr{chrnum}.covar', index=False, sep=' ', header=None)\n",
    "    print(f'wrote: chr{chrnum}.covar with {covar_df.shape[0]} samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "covar_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_ld_counts_dict = {i:0 for i in range(1,7)}\n",
    "low_ld_counts_dict = {i:0 for i in range(1,7)}\n",
    "\n",
    "for chrnum in range(1,23):\n",
    "    freqfile = f'freq.{chrnum}'\n",
    "    for Indx in range(1,7):\n",
    "        mafbinIndx = f'mafbin{Indx}'\n",
    "        mafbinfile =  f'{freqfile}.{mafbinIndx}'\n",
    "        for i in range(1, 3):\n",
    "            ldbinIndx = f'{mafbinfile}.ldbin{i}'\n",
    "            ldbin_df = pd.read_csv(ldbinIndx,sep='\\t')\n",
    "            if (i == 1):\n",
    "                high_ld_counts_dict[Indx] += ldbin_df.shape[0]\n",
    "            else:\n",
    "                low_ld_counts_dict[Indx] += ldbin_df.shape[0]\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "high ld counts per maf group: {1: 31016882, 2: 5946277, 3: 3372240, 4: 1653495, 5: 1833686, 6: 1108}\n",
      "low ld counts per maf group: {1: 31017845, 2: 5946340, 3: 3372272, 4: 1653513, 5: 1833710, 6: 1273}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>high_ld</th>\n",
       "      <th>low_ld</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>maf_bin</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>&lt;0.001</th>\n",
       "      <td>31016882</td>\n",
       "      <td>31017845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&lt;0.01</th>\n",
       "      <td>5946277</td>\n",
       "      <td>5946340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&lt;0.1</th>\n",
       "      <td>3372240</td>\n",
       "      <td>3372272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&lt;0.25</th>\n",
       "      <td>1653495</td>\n",
       "      <td>1653513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&lt;0.5</th>\n",
       "      <td>1833686</td>\n",
       "      <td>1833710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>=0.5</th>\n",
       "      <td>1108</td>\n",
       "      <td>1273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          high_ld    low_ld\n",
       "maf_bin                    \n",
       "<0.001   31016882  31017845\n",
       "<0.01     5946277   5946340\n",
       "<0.1      3372240   3372272\n",
       "<0.25     1653495   1653513\n",
       "<0.5      1833686   1833710\n",
       "=0.5         1108      1273"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f'high ld counts per maf group: {high_ld_counts_dict}')\n",
    "print(f'low ld counts per maf group: {low_ld_counts_dict}')\n",
    "high_df = pd.Series(high_ld_counts_dict).to_frame()\n",
    "low_df = pd.Series(low_ld_counts_dict).to_frame()\n",
    "high_df.columns = ['high_ld']\n",
    "low_df.columns = ['low_ld']\n",
    "ld_counts_df = high_df.merge(low_df, left_index=True, right_index=True)\n",
    "bins = ['<0.001', '<0.01', '<0.1', '<0.25', '<0.5', '=0.5']\n",
    "ld_counts_df['maf_bin'] = bins\n",
    "ld_counts_df.set_index('maf_bin', inplace=True)\n",
    "ld_counts_df.to_csv('_ld_bin_counts.txt', sep='\\t')\n",
    "ld_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--reml-no-lrt\\\n",
    "# --reml-inv-mtd 3\\\n",
    "\n",
    "gctacmd3 = f'\\\n",
    "gcta --reml\\\n",
    " --mgrm multi_GRMs.txt\\\n",
    " --pheno {idfile}\\\n",
    " --qcovar chr1.pca.eigenvec\\\n",
    " --reml-alg 2\\\n",
    " --reml-no-constrain\\\n",
    " --thread-num 60\\\n",
    " --reml-maxit 1000\\\n",
    " --out test'\n",
    "# myfunc.shell_do(gctacmd3)\n",
    "\n",
    "with open('_reml_script.swarm', 'w') as f:\n",
    "#     f.write('#!/bin/bash\\n')\n",
    "    f.write(f'{gctacmd3}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61143262\n"
     ]
    }
   ],
   "source": [
    "!swarm -f _reml_script.swarm --time=40:00:00 -g 64 -t 64 --logdir swarm --module GCTA --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df = pd.read_csv('freq.22.mafbin1',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "973767"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>jobid</th>\n",
       "      <th>jobname</th>\n",
       "      <th>state</th>\n",
       "      <th>statetime</th>\n",
       "      <th>nodelist</th>\n",
       "      <th>eval</th>\n",
       "      <th>user</th>\n",
       "      <th>submit_time</th>\n",
       "      <th>start_time</th>\n",
       "      <th>end_time</th>\n",
       "      <th>...</th>\n",
       "      <th>std_in</th>\n",
       "      <th>std_out</th>\n",
       "      <th>std_err</th>\n",
       "      <th>comment</th>\n",
       "      <th>jobidarray</th>\n",
       "      <th>jobidraw</th>\n",
       "      <th>submit_time_human</th>\n",
       "      <th>start_time_human</th>\n",
       "      <th>end_time_human</th>\n",
       "      <th>Unnamed: 47</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>61136072_0                              Biowul...</td>\n",
       "      <td>swarm</td>\n",
       "      <td>RUNNING</td>\n",
       "      <td>1594656203</td>\n",
       "      <td>cn1088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vitaled2</td>\n",
       "      <td>1594656136</td>\n",
       "      <td>1594656203</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>/dev/null</td>\n",
       "      <td>/gpfs/gsfs12/users/vitaled2/rare_variants/temp...</td>\n",
       "      <td>/gpfs/gsfs12/users/vitaled2/rare_variants/temp...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61136072_0</td>\n",
       "      <td>61136072</td>\n",
       "      <td>0x1</td>\n",
       "      <td>2020-07-13 12:02:16 EDT</td>\n",
       "      <td>2020-07-13 12:03:23 EDT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>61132199_0                              Biowul...</td>\n",
       "      <td>swarm</td>\n",
       "      <td>RUNNING</td>\n",
       "      <td>1594651119</td>\n",
       "      <td>cn3406</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vitaled2</td>\n",
       "      <td>1594651043</td>\n",
       "      <td>1594651119</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>/dev/null</td>\n",
       "      <td>/gpfs/gsfs12/users/vitaled2/rare_variants/temp...</td>\n",
       "      <td>/gpfs/gsfs12/users/vitaled2/rare_variants/temp...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61132199_0</td>\n",
       "      <td>61132199</td>\n",
       "      <td>0x1</td>\n",
       "      <td>2020-07-13 10:37:23 EDT</td>\n",
       "      <td>2020-07-13 10:38:39 EDT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>61128454                              Biowulf ...</td>\n",
       "      <td>sinteractive</td>\n",
       "      <td>RUNNING</td>\n",
       "      <td>1594646752</td>\n",
       "      <td>cn4244</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vitaled2</td>\n",
       "      <td>1594646750</td>\n",
       "      <td>1594646752</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61128454</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-07-13 09:25:50 EDT</td>\n",
       "      <td>2020-07-13 09:25:52 EDT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>61074473                              Biowulf ...</td>\n",
       "      <td>sinteractive</td>\n",
       "      <td>TIMEOUT</td>\n",
       "      <td>1594612409</td>\n",
       "      <td>cn0855</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vitaled2</td>\n",
       "      <td>1594565577</td>\n",
       "      <td>1594565580</td>\n",
       "      <td>1594612409</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61074473</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-07-12 10:52:57 EDT</td>\n",
       "      <td>2020-07-12 10:53:00 EDT</td>\n",
       "      <td>2020-07-12 23:53:29 EDT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>61078515                              Biowulf ...</td>\n",
       "      <td>_reml_script.sh</td>\n",
       "      <td>FAILED</td>\n",
       "      <td>1594572629</td>\n",
       "      <td>cn0861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vitaled2</td>\n",
       "      <td>1594572624</td>\n",
       "      <td>1594572628</td>\n",
       "      <td>1594572629</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61078515</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-07-12 12:50:24 EDT</td>\n",
       "      <td>2020-07-12 12:50:28 EDT</td>\n",
       "      <td>2020-07-12 12:50:29 EDT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>60967285                              Biowulf ...</td>\n",
       "      <td>sinteractive</td>\n",
       "      <td>TIMEOUT</td>\n",
       "      <td>1594458695</td>\n",
       "      <td>cn0927</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vitaled2</td>\n",
       "      <td>1594411836</td>\n",
       "      <td>1594411887</td>\n",
       "      <td>1594458695</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60967285</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-07-10 16:10:36 EDT</td>\n",
       "      <td>2020-07-10 16:11:27 EDT</td>\n",
       "      <td>2020-07-11 05:11:35 EDT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>60892130_4                              Biowul...</td>\n",
       "      <td>swarm</td>\n",
       "      <td>COMPLETED</td>\n",
       "      <td>1594397848</td>\n",
       "      <td>cn1948</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vitaled2</td>\n",
       "      <td>1594320771</td>\n",
       "      <td>1594320911</td>\n",
       "      <td>1594397848</td>\n",
       "      <td>...</td>\n",
       "      <td>/dev/null</td>\n",
       "      <td>/gpfs/gsfs12/users/vitaled2/rare_variants/temp...</td>\n",
       "      <td>/gpfs/gsfs12/users/vitaled2/rare_variants/temp...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60892130_4</td>\n",
       "      <td>60892356</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-07-09 14:52:51 EDT</td>\n",
       "      <td>2020-07-09 14:55:11 EDT</td>\n",
       "      <td>2020-07-10 12:17:28 EDT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>60892130_3                              Biowul...</td>\n",
       "      <td>swarm</td>\n",
       "      <td>COMPLETED</td>\n",
       "      <td>1594389998</td>\n",
       "      <td>cn1931</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vitaled2</td>\n",
       "      <td>1594320771</td>\n",
       "      <td>1594320911</td>\n",
       "      <td>1594389998</td>\n",
       "      <td>...</td>\n",
       "      <td>/dev/null</td>\n",
       "      <td>/gpfs/gsfs12/users/vitaled2/rare_variants/temp...</td>\n",
       "      <td>/gpfs/gsfs12/users/vitaled2/rare_variants/temp...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60892130_3</td>\n",
       "      <td>60892355</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-07-09 14:52:51 EDT</td>\n",
       "      <td>2020-07-09 14:55:11 EDT</td>\n",
       "      <td>2020-07-10 10:06:38 EDT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>60892130_0                              Biowul...</td>\n",
       "      <td>swarm</td>\n",
       "      <td>COMPLETED</td>\n",
       "      <td>1594368019</td>\n",
       "      <td>cn1679</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vitaled2</td>\n",
       "      <td>1594320771</td>\n",
       "      <td>1594320910</td>\n",
       "      <td>1594368019</td>\n",
       "      <td>...</td>\n",
       "      <td>/dev/null</td>\n",
       "      <td>/gpfs/gsfs12/users/vitaled2/rare_variants/temp...</td>\n",
       "      <td>/gpfs/gsfs12/users/vitaled2/rare_variants/temp...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60892130_0</td>\n",
       "      <td>60892352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-07-09 14:52:51 EDT</td>\n",
       "      <td>2020-07-09 14:55:10 EDT</td>\n",
       "      <td>2020-07-10 04:00:19 EDT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>60892130_6                              Biowul...</td>\n",
       "      <td>swarm</td>\n",
       "      <td>COMPLETED</td>\n",
       "      <td>1594367315</td>\n",
       "      <td>cn1987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vitaled2</td>\n",
       "      <td>1594320771</td>\n",
       "      <td>1594320911</td>\n",
       "      <td>1594367315</td>\n",
       "      <td>...</td>\n",
       "      <td>/dev/null</td>\n",
       "      <td>/gpfs/gsfs12/users/vitaled2/rare_variants/temp...</td>\n",
       "      <td>/gpfs/gsfs12/users/vitaled2/rare_variants/temp...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60892130_6</td>\n",
       "      <td>60892130</td>\n",
       "      <td>0x7F</td>\n",
       "      <td>2020-07-09 14:52:51 EDT</td>\n",
       "      <td>2020-07-09 14:55:11 EDT</td>\n",
       "      <td>2020-07-10 03:48:35 EDT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               jobid          jobname  \\\n",
       "0  61136072_0                              Biowul...            swarm   \n",
       "1  61132199_0                              Biowul...            swarm   \n",
       "2  61128454                              Biowulf ...     sinteractive   \n",
       "3  61074473                              Biowulf ...     sinteractive   \n",
       "4  61078515                              Biowulf ...  _reml_script.sh   \n",
       "5  60967285                              Biowulf ...     sinteractive   \n",
       "6  60892130_4                              Biowul...            swarm   \n",
       "7  60892130_3                              Biowul...            swarm   \n",
       "8  60892130_0                              Biowul...            swarm   \n",
       "9  60892130_6                              Biowul...            swarm   \n",
       "\n",
       "       state   statetime nodelist eval      user  submit_time  start_time  \\\n",
       "0    RUNNING  1594656203   cn1088  NaN  vitaled2   1594656136  1594656203   \n",
       "1    RUNNING  1594651119   cn3406  NaN  vitaled2   1594651043  1594651119   \n",
       "2    RUNNING  1594646752   cn4244  NaN  vitaled2   1594646750  1594646752   \n",
       "3    TIMEOUT  1594612409   cn0855  NaN  vitaled2   1594565577  1594565580   \n",
       "4     FAILED  1594572629   cn0861  NaN  vitaled2   1594572624  1594572628   \n",
       "5    TIMEOUT  1594458695   cn0927  NaN  vitaled2   1594411836  1594411887   \n",
       "6  COMPLETED  1594397848   cn1948  NaN  vitaled2   1594320771  1594320911   \n",
       "7  COMPLETED  1594389998   cn1931  NaN  vitaled2   1594320771  1594320911   \n",
       "8  COMPLETED  1594368019   cn1679  NaN  vitaled2   1594320771  1594320910   \n",
       "9  COMPLETED  1594367315   cn1987  NaN  vitaled2   1594320771  1594320911   \n",
       "\n",
       "     end_time  ...     std_in  \\\n",
       "0           0  ...  /dev/null   \n",
       "1           0  ...  /dev/null   \n",
       "2           0  ...        NaN   \n",
       "3  1594612409  ...        NaN   \n",
       "4  1594572629  ...        NaN   \n",
       "5  1594458695  ...        NaN   \n",
       "6  1594397848  ...  /dev/null   \n",
       "7  1594389998  ...  /dev/null   \n",
       "8  1594368019  ...  /dev/null   \n",
       "9  1594367315  ...  /dev/null   \n",
       "\n",
       "                                             std_out  \\\n",
       "0  /gpfs/gsfs12/users/vitaled2/rare_variants/temp...   \n",
       "1  /gpfs/gsfs12/users/vitaled2/rare_variants/temp...   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "5                                                NaN   \n",
       "6  /gpfs/gsfs12/users/vitaled2/rare_variants/temp...   \n",
       "7  /gpfs/gsfs12/users/vitaled2/rare_variants/temp...   \n",
       "8  /gpfs/gsfs12/users/vitaled2/rare_variants/temp...   \n",
       "9  /gpfs/gsfs12/users/vitaled2/rare_variants/temp...   \n",
       "\n",
       "                                             std_err  comment  jobidarray  \\\n",
       "0  /gpfs/gsfs12/users/vitaled2/rare_variants/temp...      NaN  61136072_0   \n",
       "1  /gpfs/gsfs12/users/vitaled2/rare_variants/temp...      NaN  61132199_0   \n",
       "2                                                NaN      NaN         NaN   \n",
       "3                                                NaN      NaN         NaN   \n",
       "4                                                NaN      NaN         NaN   \n",
       "5                                                NaN      NaN         NaN   \n",
       "6  /gpfs/gsfs12/users/vitaled2/rare_variants/temp...      NaN  60892130_4   \n",
       "7  /gpfs/gsfs12/users/vitaled2/rare_variants/temp...      NaN  60892130_3   \n",
       "8  /gpfs/gsfs12/users/vitaled2/rare_variants/temp...      NaN  60892130_0   \n",
       "9  /gpfs/gsfs12/users/vitaled2/rare_variants/temp...      NaN  60892130_6   \n",
       "\n",
       "   jobidraw  submit_time_human         start_time_human  \\\n",
       "0  61136072                0x1  2020-07-13 12:02:16 EDT   \n",
       "1  61132199                0x1  2020-07-13 10:37:23 EDT   \n",
       "2  61128454                NaN  2020-07-13 09:25:50 EDT   \n",
       "3  61074473                NaN  2020-07-12 10:52:57 EDT   \n",
       "4  61078515                NaN  2020-07-12 12:50:24 EDT   \n",
       "5  60967285                NaN  2020-07-10 16:10:36 EDT   \n",
       "6  60892356                NaN  2020-07-09 14:52:51 EDT   \n",
       "7  60892355                NaN  2020-07-09 14:52:51 EDT   \n",
       "8  60892352                NaN  2020-07-09 14:52:51 EDT   \n",
       "9  60892130               0x7F  2020-07-09 14:52:51 EDT   \n",
       "\n",
       "            end_time_human              Unnamed: 47  \n",
       "0  2020-07-13 12:03:23 EDT                      NaN  \n",
       "1  2020-07-13 10:38:39 EDT                      NaN  \n",
       "2  2020-07-13 09:25:52 EDT                      NaN  \n",
       "3  2020-07-12 10:53:00 EDT  2020-07-12 23:53:29 EDT  \n",
       "4  2020-07-12 12:50:28 EDT  2020-07-12 12:50:29 EDT  \n",
       "5  2020-07-10 16:11:27 EDT  2020-07-11 05:11:35 EDT  \n",
       "6  2020-07-09 14:55:11 EDT  2020-07-10 12:17:28 EDT  \n",
       "7  2020-07-09 14:55:11 EDT  2020-07-10 10:06:38 EDT  \n",
       "8  2020-07-09 14:55:10 EDT  2020-07-10 04:00:19 EDT  \n",
       "9  2020-07-09 14:55:11 EDT  2020-07-10 03:48:35 EDT  \n",
       "\n",
       "[10 rows x 48 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get estimates for cpu-hrs for analysis\n",
    "jobs_df = pd.read_csv('../biowulf_user_dashboard.csv')\n",
    "jobs_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "jobs_df2 = jobs_df[jobs_df.state == 'COMPLETED'][['jobid','cpu', 'cpu_min',\n",
    "       'cpu_max', 'cpu_avg', 'cpu_util', 'mem', 'mem_min', 'mem_max','elapsed_time']]\n",
    "\n",
    "jobs_df2['hrs'] = jobs_df2.elapsed_time/3600\n",
    "jobs_df2['cpu_hrs'] = jobs_df2.cpu * jobs_df2.hrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6078.294999999996"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(jobs_df2.cpu_hrs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.7",
   "language": "python",
   "name": "py3.7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
